File: app/src/main/java/fansirsqi/xposed/sesame/util/GlobalThreadPools.java
package fansirsqi.xposed.sesame.util;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * 全局线程池管理类
 */
public class GlobalThreadPools {
    private static final String TAG = "GlobalThreadPools";
    
    // 创建一个可缓存的线程池
    private static final ExecutorService THREAD_POOL = new ThreadPoolExecutor(
            5,                       // 核心线程数
            20,                      // 最大线程数
            60L,                     // 空闲线程存活时间
            TimeUnit.SECONDS,        // 时间单位
            new LinkedBlockingQueue<>(100), // 工作队列
            new ThreadFactory() {    // 线程工厂
                private final AtomicInteger threadNumber = new AtomicInteger(1);
                @Override
                public Thread newThread(Runnable r) {
                    Thread t = new Thread(r, "SesamePool-" + threadNumber.getAndIncrement());
                    if (t.isDaemon()) {
                        t.setDaemon(false);
                    }
                    if (t.getPriority() != Thread.NORM_PRIORITY) {
                        t.setPriority(Thread.NORM_PRIORITY);
                    }
                    return t;
                }
            }
    );

    /**
     * 在线程池中执行任务
     *
     * @param command 要执行的任务
     */
    public static void execute(Runnable command) {
        try {
            THREAD_POOL.execute(command);
        } catch (Exception e) {
            Log.error(TAG, "执行任务异常: " + e.getMessage());
            Log.printStackTrace(e);
            // 如果线程池执行失败，尝试在当前线程执行
            try {
                command.run();
            } catch (Exception ex) {
                Log.error(TAG, "在当前线程执行任务异常: " + ex.getMessage());
                Log.printStackTrace(ex);
            }
        }
    }

    /**
     * 使当前线程暂停指定的毫秒数。
     *
     * @param millis 毫秒数。
     */
    public static void sleep(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Log.error(TAG, "Thread sleep interrupted1 " + e.getMessage());
            Thread.currentThread().interrupt();
        } catch (Exception e1) {
            Log.printStackTrace(e1);
//            Thread.currentThread().interrupt();
            Log.error(TAG, "Thread sleep interrupted2 " + e1.getMessage());
        } catch (Throwable t) {
            Log.printStackTrace(t);
//            Thread.currentThread().interrupt();
            Log.error(TAG, "Thread sleep interrupted3 " + t.getMessage());
        }
    }

    public static void shutdownAndAwaitTermination(ExecutorService pool, long timeout, String poolName) {
        if (pool != null && !pool.isShutdown()) {
            pool.shutdown();
            try {
                if (!pool.awaitTermination(1, TimeUnit.SECONDS)) {
                    pool.shutdownNow();
                    if (!pool.awaitTermination(timeout, TimeUnit.SECONDS)) {
                        Log.runtime(TAG, "thread " + poolName + " can't close");
                    }
                }
            } catch (InterruptedException ie) {
                pool.shutdownNow();
                Thread.currentThread().interrupt();
            }
        }
    }
    
    /**
     * 关闭线程池
     */
    public static void shutdown() {
        shutdownAndAwaitTermination(THREAD_POOL, 5, "THREAD_POOL");
    }
}

File: app/src/main/java/fansirsqi/xposed/sesame/util/GlobalThreadPools.kt
package fansirsqi.xposed.sesame.util

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.Runnable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext
import kotlin.math.max
import kotlin.math.min

/**
 * @author: ghostxx
 * @date: 2025/9/17
 * @description: 全局协程调度器，用于统一管理应用内的协程，提供结构化并发和生命周期管理。
 */
object GlobalThreadPools {
    private const val TAG = "GlobalThreadPools"

    /**
     * CPU核心数
     */
    private val CPU_COUNT = Runtime.getRuntime().availableProcessors()

    /**
     * 计算密集型任务并行度
     * 根据CPU核心数动态计算，保证最佳性能。
     */
    private val COMPUTE_PARALLELISM = max(2, min(CPU_COUNT - 1, 4))

    /**
     * 全局协程作用域
     * 用于启动不绑定到特定生命周期的长寿命协程
     * 使用SupervisorJob确保一个协程的失败不会影响其他协程
     */
    private val globalScope = CoroutineScope(
        SupervisorJob() + 
        Dispatchers.Default + 
        CoroutineName("SesameGlobalScope")
    )

    /**
     * 计算密集型任务调度器
     * 适用于CPU密集型操作，如复杂计算、数据处理等
     */
    @OptIn(ExperimentalCoroutinesApi::class)
    val computeDispatcher = Dispatchers.Default.limitedParallelism(COMPUTE_PARALLELISM)

    /**
     * 在全局协程作用域中执行一个任务。
     *
     * @param block 要执行的挂起函数代码块
     * @param context 可选的协程上下文，默认使用计算调度器
     * @return 代表任务的Job对象
     */
    fun execute(
        context: CoroutineContext = computeDispatcher,
        block: suspend CoroutineScope.() -> Unit
    ): Job {
        return globalScope.launch(context) {
            try {
                block()
            } catch (_: CancellationException) {
                // 协程取消异常，正常流程，不记录
            } catch (e: Exception) {
                Log.error(TAG, "执行任务异常: ${e.message}")
                Log.printStackTrace(e)
            }
        }
    }

    /**
     * 提交一个可返回结果的任务。
     *
     * @param T 结果类型
     * @param context 可选的协程上下文，默认使用计算调度器
     * @param block 要执行的挂起函数代码块，返回类型为T
     * @return 代表任务结果的Deferred对象
     */
    fun <T> submit(
        context: CoroutineContext = computeDispatcher,
        block: suspend CoroutineScope.() -> T
    ): Deferred<T> {
        return globalScope.async(context) {
            block()
        }
    }

    /**
     * 兼容Java的执行方法
     *
     * @param command 要执行的Runnable任务
     * @return 代表任务的Job对象
     */
    @JvmOverloads
    fun execute(command: Runnable?, context: CoroutineContext = computeDispatcher): Job {
        return execute(context) {
            command?.run()
        }
    }

    /**
     * 兼容Java的提交方法
     *
     * @param task 要提交的Runnable任务
     * @return 代表任务的Deferred对象
     */
    @JvmOverloads
    fun submit(task: Runnable?, context: CoroutineContext = computeDispatcher): Deferred<Unit> {
        return submit(context) {
            task?.run()
        }
    }


    /**
     * 协程兼容的暂停方法
     * 优先使用协程delay，在非协程环境中降级到Thread.sleep
     *
     * @param millis 要暂停的毫秒数
     */
    @JvmStatic
    fun sleepCompat(millis: Long) {
        CoroutineUtils.sleepCompat(millis)
    }

}